# Problems Notebook for Applied Statistics 2025
The notebook summarizes my journey towards solving four problems set as part of the Applied Statistics module in the HDip in Computer Science and Data Analytics. The problems are:

1. Extending the Lady Tasting Tea Experiment
2. Normal distribution
3. t-Tests
4. ANOVA
 
## Standard packages used across the problems




`ipython`(https://ipython.org/)



`Numpy` is the standard package used for scientific computing in Python ([see the manual](https://numpy.org/doc/2.3/)). Numpy makes it easier to work with your data using [arrays](https://numpy.org/doc/2.3/user/absolute_beginners.html) to handle and understand your data.



`scipy`

`statsmodels`

`matplotlib` is a commmonly used package for creating visualisations of data in Python(https://matplotlib.org/stable/index.html). Visualisation of the data 

In [ ]:
import matplotlib.pyplot as plt




`pandas`enables to the analysis and manipualtion of data (https://pandas.pydata.org/docs/). It is commonly used for working with tabular data.


In [ ]:
import pandas as pd

`seaborn`


In [ ]:
import seaborn as sns

`sympy`
`nose`
`scikit-learn`
`qiskit[visualization]`
`yfinance`

In [ ]:
# Import the numpy library
import numpy as np

## Problem 1: Extending the Lady Tasting Tea
Firstly, what is the the Lady Tasting Tea Problem. This comes from the "Design of Experiments" by Ronald A fisher and states a lady can work out from tasting a cup of tea she can tell if the milk was added to the cup first or after the tea. The experiment he designed was to create eight cups of tea, half made milk first and half made tea first, and presenting them in a random order to the taster. the taster has the experiment explained so knows that four are made milk first and four are tea first. 

the null hypothesis is that she can't tell the difference between the tea making methods. There is a good explanation of [null hypothesis here](https://www.geeksforgeeks.org/maths/null-hypothesis/) if this concept is new to you. This iteration of the experiment will extend the number of cups of tea to twelve with eight made tea-first and four made milk-first.

### Import the needed tools
To complete this task we are going to need to use some mathematical functions so import `math`. Documentation about `math` is available at [https://docs.python.org/3/library/math.html](https://docs.python.org/3/library/math.html).

### Permutations and combinations
Permutations are the different ways the elements can be ordered. For example, your colleague has told you the combination for the lock on the stationery cupboard, you remember the digits are 1, 5, 7, 9, but not the order. None of the digits were repeated. The different orders of these digits are the permutations eg 1579, 1597, 9715...... The number of different ways of ordering these permutations is four factorial, often written as 4!. This is calculated by 4 * 3 * 2 * 1 = 24. It's going to take a while to get that stationery!

Combinations is a selection of a number of elements from the set. The order does not matter and the same element cannot be selected twice. If we take our stationery cuboard lock again adn all we know if no digit is repeated in the code, the how many cominations so we have? 1234, 1345, 1456 etc. The order does not matter in combinations.

How does this apply to our cups of tea experiment? First, let's import [`math`](https://docs.python.org/3/library/math.html) to help us complete the maths tasks in this experiment.

In [1]:
# Mathematic functions from the standard library
# https://docs.python.org/3/library/math.html
import math

Now, let's set our variables. The number of cups of tea in total is twelve.

In [8]:
# Number of cups of tea in total
n_cups = 12

The number of the cups made tea-first is eight.

In [9]:
n_tea_first = 8

The number of the cups made milk-first is four.

In [10]:
n_milk_first = 4

How many different ways are there for the tester to select the four cups made tea first? In the `math` library there is the [`comb`](https://docs.python.org/3/library/math.html#math.comb) function which will return the number of ways to choose a set number of items (often denoted as *k*) from *n* items without repeating items or order. Let's use the function to work out the number of ways of selecting four cups (those made milk-first) from twelve.

In [12]:
# Number of ways of selecting four cups from twelve.
n_ways = math.comb(n_cups, n_milk_first)

# Show the result
n_ways

495

There are 495 different ways of selecting the four cups made milk first. So how does `math.comb` calculate this? If we label each of the cups one to twelve and put out four coasters, the tester is asked to put the milk-first cups on the coasters. For example, cups 3,4,7,11 might be selected. the cups could be placed on the coasters in any order, so 3 4 7 11 is the same and 7 4 11 3. Once a coaster is full it reduces the number of cups to choose from by one. To calculate the number of ways ofe selecting our four cups from twelve is to mutiple the number of cups to choose from to fill each coaster. So for coaster A, there are 12 choice and coaster B 11 choices and so on.

In [13]:
# total ways of selecting 4 cups from 12
no_ways_ordered = 12 * 11 * 10 * 9

# Show the result
no_ways_ordered

11880

We don't care about the order so this reduces the number of options.

In [ ]:


# Permutations
import itertools

#Random  selections
import random

#Numberical structures and operations
import numpy as np

#Plotting
import matplotlib.pyplot as plt





Simulate this experiment using numpy by randomly shuffling the cups many times and calculating the probability of the participant correctly identifying all cups by chance. Compare your result with the original 8-cup experiment.

In your notebook, explain your simulation process clearly, report and interpret the estimated probability, and discuss whether, based on this probability, you would consider extending or relaxing the p-value threshold compared to the original design.

## Problem 2: Normal Distribution

'''Generate 100,000 samples of size 10 from the standard normal distribution. For each sample, compute the standard deviation with ddof=1 (sample SD) and with ddof=0 (population SD). Plot histograms of both sets of values on the same axes with transparency. Describe the differences you see. Explain how you expect these differences to change if the sample size is increased.'''

[Normal distribution](https://www.geeksforgeeks.org/maths/normal-distribution/) is the most common distribution of data variables. Normal  

## Problem 3: t-Tests

A type II error occurs when a test fails to reject the null hypothesis even though it is false. For each mean difference 
d=0,0.1,0.2,…,1.0, repeat the following simulation 1,000 times:

1. Draw two samples of size 100, one from the standard normal distribution and one from the normal distribution with mean d and standard deviation 1.
2. Run an independent samples t-test on the two samples, rejecting the null hypothesis if the p-value is less than 0.05.
3. Record the proportion of times the null hypothesis is not rejected.
Plot this proportion against d, and explain how the type II error rate changes as the difference in means increases.



## Problem 4: ANOVA
Generate three independent samples, each of size 30, from normal distributions with means 0, 0.5, and 1, each with standard deviation 1.

1. Perform a one-way ANOVA to test whether all three means are equal.
2. Perform three independent two-sample t-tests: samples 1 vs 2, 1 vs 3, and 2 vs 3.
Compare the conclusions.
3. Write a short note on why ANOVA is preferred over running several t-tests.

## End